hal/raspi3.c

```diff
@@ -18,23 +18,209 @@
+
+/* mail box message buffer */
+volatile unsigned int  __attribute__((aligned(16))) mbox[36];
+
+#define VIDEOCORE_MBOX  (MMIO_BASE+0x0000B880)
+#define MBOX_READ       ((volatile unsigned int*)(VIDEOCORE_MBOX+0x0))
+#define MBOX_POLL       ((volatile unsigned int*)(VIDEOCORE_MBOX+0x10))
+#define MBOX_SENDER     ((volatile unsigned int*)(VIDEOCORE_MBOX+0x14))
+#define MBOX_STATUS     ((volatile unsigned int*)(VIDEOCORE_MBOX+0x18))
+#define MBOX_CONFIG     ((volatile unsigned int*)(VIDEOCORE_MBOX+0x1C))
+#define MBOX_WRITE      ((volatile unsigned int*)(VIDEOCORE_MBOX+0x20))
+#define MBOX_RESPONSE   0x80000000
+#define MBOX_FULL       0x80000000
+#define MBOX_EMPTY      0x40000000
+
+#define MBOX_REQUEST    0
+
+/* channels */
+#define MBOX_CH_POWER   0
+#define MBOX_CH_FB      1
+#define MBOX_CH_VUART   2
+#define MBOX_CH_VCHIQ   3
+#define MBOX_CH_LEDS    4
+#define MBOX_CH_BTNS    5
+#define MBOX_CH_TOUCH   6
+#define MBOX_CH_COUNT   7
+#define MBOX_CH_PROP    8
+
+/* tags */
+#define MBOX_TAG_GETBRDVERSION  0x10002
+#define MBOX_TAG_GETSERIAL      0x10004
+#define MBOX_TAG_GET_CLOCK_RATE 0x30002
+#define MBOX_TAG_SETCLKRATE     0x38002
+#define MBOX_TAG_LAST           0
+
 /* Fixed addresses */
 extern void *kernel_addr, *update_addr, *dts_addr;
+/* Loop <delay> times  */
+static inline void delay(int32_t count)
+{
+    register unsigned int c;
+    c = count;
+    while (c--) {
+        asm volatile("nop");
+    }
+}
+/**
+ * write message to mailbox
+ */
+static void mailbox_write(uint8_t chan)
+{
+    uint32_t ch = (((uint32_t)((unsigned long)&mbox) & ~0xf)
+                | (chan & 0xf));
+
+    /* wait until mail box becomes ready to write */
+    while ((*MBOX_STATUS & MBOX_FULL) != 0) { }
+
+    /* write the address of the message to mail-box channel */
+    *MBOX_WRITE = ch;
+}
+/**
+ * read message from mailbox
+ */
+static int mailbox_read(uint8_t chan)
+{
+    uint32_t ch = (((uint32_t)((unsigned long)&mbox) & ~0xf)
+                | (chan & 0xf));
+    /* now wait for the response */
+    while(1) {
+        while ((*MBOX_STATUS & MBOX_EMPTY) != 0) { }
+        if(ch == *MBOX_READ)
+            /* is it a valid successful response */
+            return mbox[1] == MBOX_RESPONSE;
+    }
+    return 0;
+}
+
+/* UART functions for Raspberry Pi 3 UART */
+void uart_tx(char c)
+{
+    /* wait until uart channel is ready to send */
+    do{
+        asm volatile("nop");
+    } while(*UART0_FR & 0x20);
+    *UART0_DR = c;
+}
+
+char uart_read(void)
+{
+    char c;
+    /* wait until data is comming */
+    do{
+        asm volatile("nop");
+    } while(*UART0_FR & 0x10);
+    /* read it and return */
+    c = (char)(*UART0_DR);
+    /* convert carrige return to newline */
+    if (c == '\r')
+        c = '\n';
+
+    return c;
+}
+
+/**
+ * Send string to UART
+ */
+void uart_write(const char* buf, uint32_t sz) {
+    uint32_t len = sz;
+
+    while (len > 0 && *buf) {
+        /* convert newline to carrige return + newline */
```

Comment:
spelling, plus conversion from CR to CR+NL, that should be handled by the caller.

Reply:
Thanks. Agree. Carriage return code should be handled by the caller.